packages/txservice/src/provider.ts

```diff
@@ -0,0 +1,177 @@
+import { jsonifyError } from "@connext/nxtp-utils";
+import axios from "axios";
+import { BigNumber, Signer, Wallet, providers } from "ethers";
+import PriorityQueue from "p-queue";
+import { BaseLogger } from "pino";
+
+import { TransactionServiceConfig } from "./config";
+import { ChainError } from "./error";
+import { FullTransaction, MinimalTransaction, ProviderConfig } from "./types";
+
+const { JsonRpcProvider, FallbackProvider } = providers;
+
+/// Could use a more encompassing name, e.g. ChainRpcDispatch, etc
+export class ChainRpcProvider {
+  private provider: providers.JsonRpcProvider | providers.FallbackProvider;
+  private config: { gasInitialBumpPercent: number; gasPriceMinimum: BigNumber; rpcProviderMaxRetries: number; };
+  private signer: Signer;
+  private queue: PriorityQueue = new PriorityQueue({ concurrency: 1 });
+
+  public confirmationsRequired: number;
+  public chainId: number;
+
+  constructor(
+    private readonly log: BaseLogger,
+    chainId: number,
+    signer: string | Signer,
+    providerConfigs: ProviderConfig[],
+    config: TransactionServiceConfig,
+  ) {
+    const { gasInitialBumpPercent, gasPriceMinimum, rpcProviderMaxRetries } = config;
+    this.config = { gasInitialBumpPercent, gasPriceMinimum, rpcProviderMaxRetries };
+    this.confirmationsRequired = config.chainConfirmations.get(chainId) ?? config.defaultConfirmationsRequired;
+    this.chainId = chainId;
+    
+    // Register a provider for each url.
+    // Make sure all providers are ready()
+
+    if (providerConfigs.length === 1) {
+      this.provider = new JsonRpcProvider(providerConfigs[0]);
+    } else {
+      const fallbackProviderConfigs = providerConfigs.map((config) => ({
+        provider: new JsonRpcProvider(
+          {
+            url: config.url,
+            user: config.user,
+            password: config.password,
+          }
+        ),
+        priority: config.priority ?? 1,
+        weight: config.weight ?? 1,
+        stallTimeout: config.stallTimeout,
+      }));
+      // TODO: Quorum is set to 1 here, but we may want to reconfigure later. Normally it is half the sum of the weights,
+      // which might be okay in our case, but for now we have a low bar.
+      this.provider = new FallbackProvider(fallbackProviderConfigs, 1);
+    }
+
+    this.signer = typeof signer === "string" ? new Wallet(signer, this.provider) : signer.connect(this.provider);
```

Comment:
could the `signer` ever be a mnemonic?

Reply:
idk, since it's not being injected from frontend with this current architecture. But I thought, well, that could change, so better leave it

If we'd like to require it to be Signer type for now though, that's fine, lmk

Reply:
It looks like the way you've structured it is you can pass in either a signer or a private key. I think this is fine, if there's a mnemonic it can be instantiated into a signer outside of this. Also this way we can pass in the injected signer.

Reply:
Cool, I will keep then